### PR TITLE
Finished problem 4.

### PR DESCRIPTION
**Pull request by valentinawmm**

### CODE REVIEW

File: demo.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+d = {"a": 2, "b": 3}
```
Comment (W-Qiu) [inline demo.py]: Why do you submit this demo file?

File: ps3.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            return False
+        elif letter == '*':
+            for i in VOWELS:
+                wordCopy = copy(word)
+                wordCopy = wordCopy.replace('*',i)
+                if wordCopy in word_list:
+                    break
+            else:
+                return False
+    #if * can replace a vowel and consist a legal word:True
```
Comment (W-Qiu) [inline ps3.py]: What's the point of these comments?

File: ps3.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        else:
+            count_letter[letter] += 1
+    for letter in word:
+        if letter not in hand :
+            return False
+        if count_letter[letter] > hand[letter]:
+            return False
+    return True
+    
+    # option 2: minus one in hand, if reaches -1, returns False (make a copy of hand)
```
Comment (W-Qiu) [inline ps3.py]: Same, why comments here?

File: ps3.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    word_list: list of lowercase strings
+    returns: boolean
+    """
+    word = word.lower()
+    for letter in word:
+        if letter not in VOWELS and letter not in CONSONANTS and letter != '*':
+            return False
+        elif letter == '*':
+            for i in VOWELS:
+                wordCopy = copy(word)
```
Comment (W-Qiu) [inline ps3.py]: Is copy() necessary here?
Why not just wordCopy = word?

File: ps3.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+    Does not mutate hand or word_list.
+   
+    word: string
+    hand: dictionary (string -> int)
+    word_list: list of lowercase strings
+    returns: boolean
+    """
+    word = word.lower()
+    for letter in word:
+        if letter not in VOWELS and letter not in CONSONANTS and letter != '*':
```
Comment (W-Qiu) [inline ps3.py]: How about 
`
if letter not in alphabet and letter != '*'?
`
Vowels + Consonants = all 26 letters

File: ps3.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+                if wordCopy in word_list:
+                    break
+            else:
+                return False
+    #if * can replace a vowel and consist a legal word:True
+    #if * can is a consonant,that is illegal:False
+
+
+
+    # Option 1: count the number of letters in the word
```
Comment (W-Qiu) [inline ps3.py]: Remove this comment too, let your code speak for itself.

File: ps3.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+    count_letter = {}
+    for letter in word:
+        if letter not in count_letter:
```
Comment (W-Qiu) [inline ps3.py]: Only a suggestion, maybe like this:
`
counter_letter[letter] = count_letter.get(letter, 0) + 1
`

Comment (valentinawmm) [reply ps3.py]: Fixed.

**Review by W-Qiu — APPROVED** [review]

**Pull request merged**